src/instructlab/cli/model/upload.py

```diff
@@ -0,0 +1,239 @@
+import click
+
+# First Party
+from instructlab import clickext
+from instructlab.configuration import DEFAULTS
+from instructlab.defaults import DEFAULT_INDENT
+from instructlab.utils import is_model_gguf, is_model_safetensors
+
+logger = logging.getLogger(__name__)
+
+
+class ModelUploader(abc.ABC):
+    """Base class for a model uploading backend"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        destination: str,
+        release: str,
+    ) -> None:
+        self.ctx = ctx
+        self.model = model
+        self.destination = destination
+        self.release = release
+
+    @abc.abstractmethod
+    def upload(self) -> None:
+        """Uploads specified local model and stores it into destination@release"""
+
+
+class HFModelUploader(ModelUploader):
+    """Class to handle uploading safetensors and GGUF models to Hugging Face"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        destination: str,
+        release: str,
+        hf_token: str,
+    ) -> None:
+        super().__init__(
+            ctx=ctx,
+            model=model,
+            destination=destination,
+            release=release,
+        )
+        self.hf_token = (
+            hf_token.strip()
+        )  # Remove trailing whitespaces from the token if they exist
+        self.local_model_path = ""
+
+    def upload(self):
+        """
+        Upload specified model to Hugging Face
+        """
+
+        # HF token check
+        if len(self.hf_token) == 0:
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to upload HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
+
+        click.echo(
+            f"Uploading model to Hugging Face:\n{DEFAULT_INDENT}Model: {self.model}\n{DEFAULT_INDENT}Destination: {self.destination}@{self.release}\n"
+        )
+
+        # set logger
+        if self.ctx.obj is not None:
+            hf_logging.set_verbosity(self.ctx.obj.config.general.log_level.upper())
+
+        # determine if model is path or name - look in checkpoints dir if the latter
+        if os.path.exists(self.model):
+            self.local_model_path = Path(self.model)
+            self.model = self.model.split("/")[-1]
+        else:
+            self.local_model_path = os.path.join(DEFAULTS.CHECKPOINTS_DIR, self.model)
+            # throw an error if the built path is invalid
+            if not os.path.exists(self.local_model_path):
+                click.secho(
+                    f"Couldn't find model at {self.local_model_path} - are you sure it exists?",
+                    fg="red",
+                )
+                raise click.exceptions.Exit(1)
+
+        # upload either safetensors or gguf, if valid
+        if is_model_safetensors(self.local_model_path):
```

Comment:
we should support .bin, I think there is a separate function for this (list supports this, as does download I think)

Reply:
this is another one I think should be addressed before merge besides the cli split mentioned in the review comment

Reply:
Do we support `.bin` downloads?

Reply:
yes https://github.com/instructlab/instructlab/pull/2283

all trained models are also in .bin format.

Reply:
Ack, I will make this change!